src/video_core/control/scheduler.cpp

```diff
@@ -1,32 +1,204 @@
 // SPDX-FileCopyrightText: 2021 yuzu Emulator Project
 // SPDX-License-Identifier: GPL-3.0-or-later
 
+#include <atomic>
+#include <deque>
+#include <map>
 #include <memory>
+#include <mutex>
+#include <unordered_map>
 
 #include "common/assert.h"
-#include "video_core/control/channel_state.h"
+#include "common/fiber.h"
 #include "video_core/control/scheduler.h"
+#include "video_core/dma_pusher.h"
 #include "video_core/gpu.h"
 
 namespace Tegra::Control {
-Scheduler::Scheduler(GPU& gpu_) : gpu{gpu_} {}
+
+struct GPFifoContext {
+    bool is_active;
+    bool is_running;
+    std::shared_ptr<Common::Fiber> context;
+    std::deque<CommandList> pending_work;
+    std::mutex guard;
+    s32 bind_id;
+    std::shared_ptr<ChannelState> info;
+    size_t yield_count;
+    size_t scheduled_count;
+};
+
+struct Scheduler::SchedulerImpl {
+    // Fifos
+    std::map<u32, std::list<size_t>, std::greater<u32>> schedule_priority_queue;
+    std::unordered_map<s32, size_t> channel_gpfifo_ids;
+    std::deque<GPFifoContext> gpfifos;
+    std::deque<size_t> free_fifos;
+
+    // Scheduling
+    std::mutex scheduling_guard;
+    std::shared_ptr<Common::Fiber> master_control;
+    bool must_reschedule{};
+    GPFifoContext* current_fifo{};
+};
+
+Scheduler::Scheduler(GPU& gpu_) : gpu{gpu_} {
+    impl = std::make_unique<SchedulerImpl>();
+}
 
 Scheduler::~Scheduler() = default;
 
+void Scheduler::Init() {
+    impl->master_control = Common::Fiber::ThreadToFiber();
+}
+
+void Scheduler::Resume() {
+    bool pending_work;
+    do {
+        pending_work = false;
+        {
+            std::unique_lock lk(impl->scheduling_guard);
+            impl->current_fifo = nullptr;
+            auto it = impl->schedule_priority_queue.begin();
+            while (it != impl->schedule_priority_queue.end()) {
+                pending_work = ScheduleLevel(it->second);
+                if (pending_work) {
+                    break;
+                }
+                it = std::next(it);
+            }
+            if (pending_work) {
+                impl->must_reschedule = false;
+            }
+        }
+        if (impl->current_fifo) {
+            impl->current_fifo->scheduled_count++;
+            Common::Fiber::YieldTo(impl->master_control, *impl->current_fifo->context);
+        }
+    } while (pending_work);
+}
```

Comment:
Extract the update logic to a new method
```cpp
bool Scheduler::UpdateHighestPriorityChannel() {
    std::scoped_lock lk(impl->scheduling_guard);

    // Clear needs to schedule state.
    impl->must_reschedule = false;

    // By default, we don't have a channel to schedule.
    impl->current_fifo = nullptr;

    // Check each level to see if we can schedule.
    for (auto& level : impl->schedule_priority_queue) {
        if (ScheduleLevel(level)) {
            return true;
        }
    }

    // Nothing to schedule.
    return false;
}
```
Then this becomes as simple as
```suggestion
void Scheduler::Resume() {
    while (UpdateHighestPriorityChannel()) {
        impl->current_fifo->scheduled_count++;
        Common::Fiber::YieldTo(impl->master_control, *impl->current_fifo->context);
    }
}
```